### PR TITLE
FOGL-8759 Password Policy changes

### PR DESCRIPTION
**Pull request by ashish-jabble**

### CODE REVIEW

**Review by MarkRiddoch — APPROVED** [review]

File: docs/building_fledge/06_testing.rst [hunk 1]
```diff
@@ -455,49 +455,78 @@ If you are curious to see which categories are available in Fledge, simply type:
...
+          "description": "Service Monitor",
+          "displayName": "Service Monitor"
+        },
+        {
+          "key": "rest_api",
+          "description": "Fledge Admin and User REST API",
+          "displayName": "Admin API"
+        },
+        {
+          "key": "password",
```
Comment (praveen-garg) [inline docs/building_fledge/06_testing.rst:490]: key should be passwordPolicy. This is not a password

File: docs/securing_fledge.rst [hunk 2]
```diff
@@ -128,6 +129,27 @@ Towards the bottom of this profile display the *change password* option appears.
 
 This popup can be used to change your password. On successfully changing your password you will be logged out of the user interface and will be required to log back in using this new password.
 
+Password Policy
+---------------
+
+Fledge provides different policies to control the passwords. The following options are currently available:
```
Comment (praveen-garg) [inline docs/securing_fledge.rst:135]: Which passwords?

File: docs/rest_api_guide/03_RESTadmin.rst [hunk 3]
```diff
@@ -393,98 +422,90 @@ The response payload is a set of configuration items within the category, each i
...
+      "authMethod": {
+        "description": "Authentication method",
+        "type": "enumeration",
+        "options": [
+          "any",
+          "password",
+          "certificate"
+        ],
+        "default": "any",
+        "displayName": "Authentication method",
```
Comment (praveen-garg) [inline docs/rest_api_guide/03_RESTadmin.rst:480]: m should be M of method, to be consistent.

File: python/fledge/services/core/api/auth.py [hunk 4]
```diff
@@ -582,10 +578,11 @@ async def update_password(request):
         raise web.HTTPBadRequest(reason=msg)
 
     if new_password and not isinstance(new_password, str):
-        raise web.HTTPBadRequest(reason=PASSWORD_ERROR_MSG)
-    if new_password and not re.match(PASSWORD_REGEX_PATTERN, new_password):
-        raise web.HTTPBadRequest(reason=PASSWORD_ERROR_MSG)
-
+        err_msg = "New password should be in string format."
```
Comment (praveen-garg) [inline python/fledge/services/core/api/auth.py:581]: should be a valid string

not string format

File: python/fledge/services/core/api/auth.py [hunk 5]
```diff
@@ -816,3 +814,41 @@ def has_admin_permissions(request):
...
+
+    message = ""
+    storage_client = connect.get_storage_async()
+    cfg_mgr = ConfigurationManager(storage_client)
+    category = await cfg_mgr.get_category_all_items('password')
+    policy = category['policy']['value']
+    min_chars = category['length']['value']
+    max_chars = category['length']['maximum']
+    if len(password) < int(min_chars):
+        message = "Password length is minimum of {} characters.".format(min_chars)
```
Comment (praveen-garg) [inline python/fledge/services/core/api/auth.py:831]: not of

File: python/fledge/services/core/api/auth.py [hunk 6]
```diff
@@ -816,3 +814,41 @@ def has_admin_permissions(request):
...
+            if not mixed_case:
+                message = "Password must contain upper and lower case letters."
+        elif policy == 'Mixed case and numeric':
+            mixed_numeric_case = has_lower and has_upper and has_numeric
+            if not mixed_numeric_case:
+                message = "Password must contain upper, lower case, uppercase and numeric values."
+        elif policy == 'Mixed case, numeric and special characters':
+            mixed_numeric_special_case = has_lower and has_upper and has_numeric and has_special
+            if not mixed_numeric_special_case:
+                message = "Password must contain atleast one upper and lower case letter, numeric and special characters."
```
Comment (praveen-garg) [inline python/fledge/services/core/api/auth.py:850]: one upper case, lower case,

File: python/fledge/services/core/server.py [hunk 7]
```diff
@@ -487,6 +480,44 @@ async def rest_api_config(cls):
...
             raise
 
+    @classmethod
+    async def password_config(cls):
+        try:
+            config = {
+                'policy': {
+                    'description': 'Password policy',
+                    'type': 'enumeration',
+                    'options': ['Any characters', 'Mixed case Alphabetic', 'Mixed case and numeric', 'Mixed case, numeric and special characters'],
```
Comment (praveen-garg) [inline python/fledge/services/core/server.py:490]: alphabetic

File: python/fledge/services/core/user_model.py [hunk 8]
```diff
@@ -336,7 +336,7 @@ async def login(cls, username, password, host):
             # check password change configuration
             storage_client = connect.get_storage_async()
             cfg_mgr = ConfigurationManager(storage_client)
-            category_item = await cfg_mgr.get_category_item('rest_api', 'passwordChange')
+            category_item = await cfg_mgr.get_category_item('password', 'expiration')
```
Comment (praveen-garg) [inline python/fledge/services/core/user_model.py:339]: parent of these should be rest_api for Admin/ other users?

**Pull request merged**